Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

# Importing and reading dataset

In [0]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

In [66]:
df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

# engineering features

In [0]:
# total perks
perks = ['elevator', 'cats_allowed', 'hardwood_floors', 'dogs_allowed', 'doorman', 'dishwasher', 'no_fee', 'laundry_in_building', 'fitness_center', 'pre-war', 'laundry_in_unit', 'roof_deck', 
         'outdoor_space', 'dining_room', 'high_speed_internet', 'balcony', 'swimming_pool', 'new_construction', 'terrace', 'exclusive', 'loft', 'garden_patio', 'wheelchair_access', 'common_outdoor_space']

In [0]:
# total rooms
rooms = ['bathrooms', 'bedrooms']

In [69]:
df.shape

(48817, 34)

In [0]:
df = df.reset_index(drop=True)

In [0]:
def sum_line(df, columns):
  total= []
  for x in range(len(df)):
    t = sum(df[columns].loc[x])
    total.append(t)

  return total

In [72]:
total_rooms = sum_line(df, rooms)
total_rooms[0:5]

[4.5, 3.0, 2.0, 2.0, 5.0]

In [73]:
total_perks = sum_line(df, perks)
total_perks[0:5]

[0, 5, 3, 2, 1]

In [0]:
df['total_rooms'] = total_rooms
df['total_perks'] = total_perks

In [82]:
df.head()

bathrooms  bedrooms  ... total_rooms total_perks
0        1.5         3  ...         4.5           0
1        1.0         2  ...         3.0           5
2        1.0         1  ...         2.0           3
3        1.0         1  ...         2.0           2
4        1.0         4  ...         5.0           1

[5 rows x 36 columns]

# spliting data into train and test

In [0]:
df['created'] = df['created'].apply(lambda x: pd.to_datetime(x))

In [0]:
# training data April & May
time_train = (df['created'].dt.month == 4) | (df['created'].dt.month == 5)
train = df[time_train]

time_test = df['created'].dt.month == 6
test = df[time_test]

# Linear Regression 2 features

In [0]:
from sklearn.linear_model import LinearRegression

In [0]:
model = LinearRegression(fit_intercept=True)

In [0]:
target = 'price'
y_train = train[target]
y_test = test[target]

In [0]:
features = ['total_rooms', 'total_perks']
X_train = train[features]
X_test = test[features]

In [105]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [108]:
df.sample(1)

bathrooms  bedrooms  ... total_rooms total_perks
37387        2.0         2  ...         4.0           6

[1 rows x 36 columns]

In [110]:
model.predict([[4, 6]])

array([4680.60723428])

# model's coefficients and intercept

In [111]:
model.coef_

array([767.95662272, 101.76435155])

In [101]:
model.intercept_

0.0

In this model each room adds $767.96 and each perk adds $101.76. The intercept is zero because I have the model to fit_intercept=True.

# Regression metrics RMSE, MAE, and  R2

## Root Mean Squared Error

In [0]:
from sklearn.metrics import mean_squared_error

In [146]:
y_pred = model.predict(X_train)
rmse = np.sqrt(mean_squared_error(y_train, y_pred))

print(f'Train Root Mean Squared Error (Apartment Price April-May): {rmse:.2f}')

Train Root Mean Squared Error (Apartment Price April-May): 1295.71


In [147]:
y_pred = model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print(f'Test Root Mean Squared Error (Apartment Price June): {rmse:.2f}')

Test Root Mean Squared Error (Apartment Price June): 1293.23


## mean_absolute_error

In [0]:
from sklearn.metrics import mean_absolute_error

In [142]:
y_pred = model.predict(X_train)
mae = mean_absolute_error(y_train, y_pred)

print(f'Train Mean_Absolute_Error (Apartment Price April-May): ${mae:.2f}')

Train Mean_Absolute_Error (Apartment Price April-May): $856.62


In [143]:
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)

print(f'Test Mean_Absolute_Error (Apartment Price June): ${mae:.2f}')

Test Mean_Absolute_Error (Apartment Price June): $868.00


## R2

In [0]:
from sklearn.metrics import r2_score

In [140]:
y_pred = model.predict(X_train)
r2 = r2_score(y_train, y_pred)

print(f'Train R^2 (Apartment Price April-May): {r2:.4f}')

Train R^2 (Apartment Price April-May): 0.4593


In [141]:
y_pred = model.predict(X_test)
r2 = r2_score(y_test, y_pred)

print(f'Train R^2 (Apartment Price June): {r2:.4f}')

Train R^2 (Apartment Price June): 0.4619
